File: go/src/leetcode/offer/Offer31.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/4 23:27
 */

func validateStackSequences(pushed []int, popped []int) bool {
	stack := make([]int, 0)
	for i := 0; i < len(pushed); i++ {
		stack = append(stack, pushed[i])
		for len(popped) > 0 && len(stack) > 0 && stack[len(stack)-1] == popped[0] {
			stack = stack[:len(stack)-1]
			popped = popped[1:]
		}
	}
	return len(popped) == 0
}

File: java/src/leetcode/offer/Offer31.java
package leetcode.offer;

import java.util.Stack;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/9/12 23:40
 */
public class Offer31 {

    public boolean validateStackSequences(int[] pushed, int[] popped) {
        int n = pushed.length;
        Stack<Integer> stack = new Stack<>();
        int index = 0;
        for (int num : pushed) {
            //按照压栈序列的顺序执行。
            stack.push(num);
            // 每次入栈后，循环判断 “栈顶元素 === 弹出序列的当前元素” 是否成立，将符合弹出序列顺序的栈顶元素全部弹出。
            while (!stack.isEmpty() && stack.peek() == popped[index]) {
                stack.pop();
                index++;
            }
        }
        //若 stack 为空，则此弹出序列合法。
        return index == n;
    }
}
